src/ingestor.js

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2023 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import { timeoutSignal, fetch as originalFetch } from '@adobe/fetch';
+import { forEachLimit } from 'async';
+import fetchBuilder from 'fetch-retry';
+
+// eslint-disable-next-line no-unused-vars
+import * as extractors from './extractors.js';
+
+const fetch = fetchBuilder(originalFetch);
+
+/**
+ * @typedef IngestorConfig the confiuration for the ingestor client
+ * @property {string} url the URL for calling the ingestor
+ * @property {string} apiKey the API Key used to call the ingestor
+ * @property {string} jobId the id of the current job
+ */
+
+/**
+ * The ingestor client sends asset data to the Content Lake ingestion service to be ingested
+ */
+export class IngestorClient {
+  #config;
+
+  #log;
+
+  /**
+   * Construct a new ingestor
+   * @param {IngestorConfig} config the configuration for the ingestor
+   */
+  constructor(config) {
+    this.#config = config;
+    this.#log = console;
+  }
+
+  withLog(logger) {
+    this.log = logger;
+    return this;
+  }
+
+  /**
+   * Submits the data and binary reference for ingestion
+   * @param {extractors.AssetData} data the asset data to ingest
+   * @param {extractors.BinaryRequest} binaryRequest the reference to the binary to ingest
+   */
+  async submit(data, binaryRequest) {
+    const signal = timeoutSignal(5000);
+    const start = Date.now();
+    const body = {
+      jobId: this.#config.jobId,
+      data,
+      binaryRequest,
+    };
+    this.#log.debug('Submitting for ingestion', body);
+    const res = await fetch(this.#config.url, {
+      headers: {
+        'X-API-Key': this.#config.apiKey,
+      },
+      method: 'POST',
+      signal,
+      body: JSON.stringify(body),
+      retryDelay: (attempt, _err, response) => {
+        let delay;
+        if (response.headers.has('Retry-After')) {
+          delay = response.headers.get('Retry-After') * 1000;
+        } else {
+          delay = attempt ** 2 * 1000;
+        }
+        this.#log.info('Retrying request', {
+          attempt,
+          delay,
+          status: `${response.status}: ${response.statusText}`,
+          url: response.url,
```

Comment:
I know NUI has been big on being very careful about the information it logs. Is anything being logged here potentially sensitive information? i.e. tokens in URLs, keys in body responses, etc.

Reply:
+1 we probably need a redactor hooked into the logs, or maybe the helix universal logger already has one? (here is what we did for nui: https://github.com/adobe/asset-compute-commons/blob/master/lib/log-utils.js)

What is response.url in this case? (If its just the ingestor lambda function url that should be fine)

Reply:
It doesn't. For now, I've made an improvement to destructure and restructure the data objects to just include the identifiers.